example/eblob_backend.cpp

```diff
@@ -747,95 +748,206 @@ struct iterated_key_info {
+typedef std::function<int (std::shared_ptr<iterated_key_info> info)> iterator_server_send_callback;
 
-static iterator_callback make_iterator_server_send_callback(eblob_backend_config *c, dnet_net_state *st,
-                                                            dnet_cmd *cmd,
-                                                            ioremap::elliptics::dnet_server_send_request &request,
-                                                            uint64_t iterator_id,
-                                                            uint64_t &counter) {
+/*
+ * congestion_control_monitor is used for limiting amount of data being sent to a remote backend.
+ */
+class congestion_control_monitor
+{
+public:
+	/*!
+	 * Constructor: initializes internal state properly.
+	 \a start_batch_size - maximum amount of data being sent as a first batch
+	 \a batch_timeout_msec - batch timeout in milliseconds
+	 */
+	congestion_control_monitor(uint64_t start_batch_size, int batch_timeout_msec)
+	: m_start_batch_size(start_batch_size), m_batch_size(m_start_batch_size),
+	 m_bytes_pending(0), m_bytes_processed(0),
+	 m_batch_timeout_msec(batch_timeout_msec),
+	 m_need_start_time(true) {
+	}
+
+	/*!
+	 * Waits until batch had any space available, then increments
+	 * number of pending bytes in the batch.
+	 */
+	void add_bytes(uint64_t bytes) {
+		std::unique_lock<std::mutex> lock{m_mutex};
+
+		while (is_batch_full()) {
+			m_cond.wait(lock);
+		}
+		m_bytes_pending += bytes;
+
+		if (m_need_start_time) {
+			gettimeofday(&m_batch_start_time, nullptr);
+			m_need_start_time = false;
+		}
+	}
+
+	/*!
+	 * Decrements number of pending bytes in the batch and measures
+	 * elapsed time of processing of the batch for either increasing or decreasing size of
+	 * the batch.
+	 */
+	void remove_bytes(uint64_t bytes) {
+		std::unique_lock<std::mutex> lock{m_mutex};
+
+		m_bytes_processed += bytes;
+
+		if (m_bytes_processed >= m_batch_size) {
+			m_bytes_processed -= m_batch_size;
+
+			struct timeval tv;
+			gettimeofday(&tv, nullptr);
+			const int64_t elapsed = static_cast<int64_t>((tv.tv_sec - m_batch_start_time.tv_sec) * 1000 +
+								     (tv.tv_usec - m_batch_start_time.tv_usec ) / 1000);
+			if (elapsed < m_batch_timeout_msec) {
+				m_batch_size *= 2;
+			} else {
+				m_batch_size = std::min(m_batch_size / 2, m_start_batch_size);
+			}
+
+			m_need_start_time = true;
+		}
+
+		m_bytes_pending -= bytes;
+
+		if (!is_batch_full()) {
+			m_cond.notify_one();
+		}
+	}
+
+	/*!
+	 * Waits until all bytes are processed
+	 */
+	void wait_completion() {
+		std::unique_lock<std::mutex> lock{m_mutex};
+		while (m_bytes_pending > 0) {
+			m_cond.wait(lock);
+		}
+	}
+
+private:
+	bool is_batch_full() const {
+		return m_bytes_pending >= m_batch_size;
+	}
+
+public:
+	/*
+	 * This constants can be used in constructor of this class.
+	 */
+	static const uint64_t DEFAULT_START_BATCH_SIZE = 1024 * 1024;
```

Comment:
Please, add description to both constants. Why does default batch size is set to 1KB?